### ISSUE TITLE
Problem with queue_timeout option

### ISSUE THREAD

**Issue by mbirtwell** [GrahamDumpleton/mod_wsgi#456]
When a queue timeout is configured if a request overruns it's queue time then the attempt to process that request is aborted but no response is sent, the request seems to continue to hold the process/thread busy until a further timeout occurs. This means that if there's one request to process that takes a long time then requests can get queued up and then the all start hitting the queue timeout and then the process timeout. e.g.:

This is a log of just 2 requests coming in the first takes longer than the queue timeout to process. The second is therefore queued for that length of time and generates a message saying "Queue timeout expired", there's no response generated at this point instead the response is generated when the servers timeout is reached.

```
Server URL         : http://localhost/
Server Root        : /tmp/mod_wsgi-localhost:80:0
Server Conf        : /tmp/mod_wsgi-localhost:80:0/httpd.conf
Error Log File     : /dev/stderr (warn)
Access Log File    : /dev/stdout
Startup Log File   : /dev/stderr
Request Capacity   : 1 (1 process * 1 threads)
Request Timeout    : 60 (seconds)
Startup Timeout    : 15 (seconds)
Queue Backlog      : 100 (connections)
Queue Timeout      : 10 (seconds)
Server Capacity    : 20 (event/worker), 20 (prefork)
Server Backlog     : 500 (connections)
Locale Setting     : en_US.UTF-8
[Thu Aug 15 11:53:15.733210 2019] [mpm_event:notice] [pid 15:tid 140623012833024] AH00489: Apache/2.4.41 (Unix) mod_wsgi/4.6.7 Python/2.7 configured -- resuming normal operations
[Thu Aug 15 11:53:15.733331 2019] [core:notice] [pid 15:tid 140623012833024] AH00094: Command line: 'httpd (mod_wsgi-express) -f /tmp/mod_wsgi-localhost:80:0/httpd.conf -E /dev/stderr -D MOD_WSGI_ACCESS_LOG -D MOD_WSGI_MULTIPROCESS -D MOD_WSGI_MPM_ENABLE_EVENT_MODULE -D MOD_WSGI_MPM_EXISTS_EVENT_MODULE -D MOD_WSGI_MPM_EXISTS_WORKER_MODULE -D MOD_WSGI_MPM_EXISTS_PREFORK_MODULE -D FOREGROUND'
[Thu Aug 15 11:53:42.328881 2019] [wsgi:error] [pid 17:tid 140622829508352] [remote 172.17.0.1:52090] mod_wsgi (pid=17): Queue timeout expired for WSGI daemon process 'localhost:80'.
172.17.0.1 - - [15/Aug/2019:11:53:22 +0000] "GET /delay/20/long HTTP/1.1" 200 12
172.17.0.1 - - [15/Aug/2019:11:53:22 +0000] "GET /delay/2/short HTTP/1.1" 504 247
[Thu Aug 15 11:54:22.858777 2019] [wsgi:error] [pid 35:tid 140623012558592] [client 172.17.0.1:52090] Timeout when reading response headers from daemon process 'localhost:80': /tmp/mod_wsgi-localhost:80:0/htdocs/delay
```
The log from my client program (BST a.k.a UTC+1):
```
2019-08-15 12:53:23,713 Starting request long
2019-08-15 12:53:24,215 Starting request short
2019-08-15 12:53:43,739 Done request long status: 200 in: 20.02531361579895
2019-08-15 12:54:24,269 Done request short status: 504 in: 60.054267168045044
```

As you can see this has a queue timeout of 10s. The queue timeout log line occurs 18s after the request is sent. And the response is finally sent after 60s. 

Ideally I think the response should be sent for a request when the timeout is reached. But looking at the code that could be tricky. Sending the response at the point that the log line is generated seems like an excellent second best, but I can't see why that wouldn't happen. 

I've put the code I've used to reproduce this in https://github.com/mbirtwell/reproduce_mod_wsgi_queue_timeout_bug, I've reproduced this behaviour with mod_wsgi 4.4.8, 4.5.15 on my systems and based on your docker images using 4.5.18 and 4.6.7. I also reproduced this with multiple threads or processes, and the behaviour was much the same. The run2 directories it that github provides an example of things getting further behind. The run3 directories show a slight variation of this where by setting the request-timeout to be less than the server level timeout I've managed to provoke a 500 error from mod_wsgi.

**Comment by mbirtwell** [comment]
I figured I'd have another go at try to get a fix for this myself. I've got as far as discovering that `wsgi_hook_daemon_handler` appears to be hanging in `ap_finalize_request_protocol` when `queue_timeout_occurred` is set. Any ideas why that would be?

**Comment by GrahamDumpleton** [comment]
Sorry for delay on commenting. Been super busy the last week and this quickly sank in the inbox and overlapped. I'll try and have a look at it over the next few days when travelling and at conference.

**Comment by GrahamDumpleton** [comment]
The code uses:

```
    if (wsgi_daemon_process->group->queue_timeout) {
        if (config->request_start) {
            apr_time_t queue_time = 0;

            queue_time = config->daemon_start - config->request_start;

            if (queue_time > wsgi_daemon_process->group->queue_timeout) {
                queue_timeout_occurred = 1;

                r->status_line = "200 Timeout";

                ap_log_rerror(APLOG_MARK, APLOG_ERR, 0, r,
                             "mod_wsgi (pid=%d): Queue timeout expired "
                             "for WSGI daemon process '%s'.", getpid(),
                             wsgi_daemon_process->group->name);
            }
        }
    }
```

It like should be:

```
    if (wsgi_daemon_process->group->queue_timeout) {
        if (config->request_start) {
            apr_time_t queue_time = 0;

            queue_time = config->daemon_start - config->request_start;

            if (queue_time > wsgi_daemon_process->group->queue_timeout) {
                queue_timeout_occurred = 1;

                r->status = HTTP_INTERNAL_SERVER_ERROR;
                r->status_line = "200 Timeout";

                ap_log_rerror(APLOG_MARK, APLOG_ERR, 0, r,
                             "mod_wsgi (pid=%d): Queue timeout expired "
                             "for WSGI daemon process '%s'.", getpid(),
                             wsgi_daemon_process->group->name);
            }
        }
    }
```

That is, also set `r->status`, which will force immediate closure of the socket connection.

**Comment by mbirtwell** [comment]
Sorry, for the slow response. That seems to work for me. Thank you very much.

**Comment by GrahamDumpleton** [comment]
Thanks for reminding me about this. Been so busy the past month that had totally forgotten hadn't even checked in the change.

**Comment by GrahamDumpleton** [comment]
I think I released this back in 4.6.8, so closing issue.
